### ISSUE TITLE
API Proposal: Add `IDeviceContext` to events that hold `Graphics`

### ISSUE THREAD

**Issue by JeremyKuhne** [dotnet/winforms#3570]
## Proposal

Add `IDeviceContext` to all Windows Forms `EventArgs` that wrap a `System.Drawing.Graphics`.

- `PaintEventArgs`
- `DrawItemEventArgs`
- `DataGridViewRowPostPaintEventArgs`
- `DrawListViewItemEventArgs`
- `DrawToolTipEventArgs`
- `DrawTreeNodeEventArgs`
- `ToolStripArrowRenderEventArgs`
- `ToolStripContentPanelRenderEventArgs`
- `ToolStripItemRenderEventArgs`
- `ToolStripPanelRenderEventArgs`
- `ToolStripRenderEventArgs`

``` C#
namespace System.Windows.Forms
{
    public class PaintEventArgs : EventArgs, IDisposable, IDeviceContext
    {
        IntPtr IDeviceContext.GetHdc() => Graphics?.GetHdc() ?? IntPtr.Zero;
        void IDeviceContext.ReleaseHdc() => Graphics?.ReleaseHdc();
    }

    public class DrawItemEventArgs: EventArgs, IDeviceContext
    {
        IntPtr IDeviceContext.GetHdc() => Graphics?.GetHdc() ?? IntPtr.Zero;
        void IDeviceContext.ReleaseHdc() => Graphics?.ReleaseHdc();

        void IDisposable.Dispose()
        {
            // IDeviceContext derives from IDisposable. We will not dispose of the contained Graphics
            // as we historically did not take ownership of it.
        }
    }
}
```

## Justification

Windows Forms was designed around GDI+, which `System.Drawing` is a wrapper around. `Graphics` was the primary abstraction. Performance of GDI+ wasn't sufficient and internally a number of code paths were changed to utilize the GDI `HDC`. `Graphics` exposes the `HDC` via the `IDeviceContext` interface, which WinForms uses internally and also exposes on public APIs (see [`VisualStyleRenderer`](https://apisof.net/catalog/System.Windows.Forms.VisualStyles.VisualStyleRenderer) and [`TextRenderer`](https://apisof.net/catalog/System.Windows.Forms.TextRenderer)).

Exposing `IDeviceContext` allows us to stop unwrapping event args and allows us to extract the `HDC` with internal optimizations:

- Internal state info on the event args allows us to know that we don't need to apply state or create a `Graphics` object if we haven't already.
- It also allows end users to pass event args back into `IDeviceContext` APIs without retrieving `Graphics`, which can create `Graphics` or force us to apply it's (likely unchanged) state.

## Implementation Details

We'll explicitly implement `IDeviceContext` to avoid having the methods directly visible on the event args. The only real benefit is treating the event args as an `IDeviceContext` object.

On event args where we don't already have `IDisposable` we'll explicitly define `IDisposable.Dispose()` and we will **not** dispose passed in arguments (notably `Graphics`).

**Comment by JeremyKuhne** [comment]
How the smuggling works:

``` C#
namespace System.Windows.Forms
{
    public class PaintEventArgs : EventArgs, IDeviceContext, IInternalInterface {}

    public class SomeControl : Control
    {
         protected override void OnPaint(PaintEventArgs e)
         {
              RendererHelper.DrawThing(e);
         }
    }

    internal static class RendererHelper
    {
        internal static void DrawThing(IDeviceContext context)
        {
             if (context is IInternalInterface internals) ...
        }
    }
}
```

**Comment by terrajobst** [comment]
Looks good as proposed.

**Comment by weltkante** [comment]
> We should implement IDisposable following the dispose pattern

Even if you make it public instead of explicit it still should not dispose the contained `Graphics` instance as the `EventArgs` instance doesn't own the Graphics.

**Comment by JeremyKuhne** [comment]
> Even if you make it public instead of explicit it still should not dispose the contained Graphics instance as the EventArgs instance doesn't own the Graphics instance.

Yes, and I comment on that in the original description. We never dispose passed in `Graphics` objects.

**Comment by RussKie** [comment]
Resolved in #3553
